File: 10-1.rb
#!/usr/bin/env ruby

class CPU
  attr_accessor :cycle, :strength, :x

  def initialize
    @x = 1
    @cycle = 0
    @strength = 0
    @next_sample = 20
  end

  def advance_timer(cycles)
    if @cycle + cycles >= @next_sample
      @strength += @next_sample * @x
      @next_sample += 40
    end
    @cycle += cycles
  end

  def addx(v)
    advance_timer 2
    @x += v
  end

  def noop(_)
    advance_timer 1
  end

  def to_s
    "<#{self.class}: X: #{@x}, cycle: #{@cycle}, strength: #{@strength}>"
  end
end

instructions = File.read('10.input').lines.map(&:strip)

instruction_format = /(?<inst>addx|noop) ?(?<oper>-?[[:digit:]]+)?/

cpu = CPU.new

instructions.each do |insn|
  instruction_format.match(insn) { |m| cpu.send(m['inst'], m['oper'].to_i) }
end

print cpu.strength, "\n"

File: 10-1.py
#!/usr/bin/env python

import re


class CPU:
    def __init__(self):
        self.x = 1
        self.cycle = 0
        self.strength = 0
        self.next_sample = 20

    def advance_timer(self, cycles):
        if self.cycle + cycles >= self.next_sample:
            self.strength += self.next_sample * self.x
            self.next_sample += 40

        self.cycle += cycles

    def addx(self, v):
        self.advance_timer(2)
        self.x += int(v)

    def noop(self, _):
        self.advance_timer(1)

    def __str__(self):
        return f"<{self.__class__.__name__}: X: {self.x}, cycle: {self.cycle}, strength: {self.strength}>"


instructions = [insn.strip() for insn in open('10.input').readlines()]

instruction_format = re.compile("(?P<inst>addx|noop) ?(?P<oper>-?[0-9]+)?")

cpu = CPU()

for insn in instructions:
    m = instruction_format.match(insn)
    fun = getattr(cpu, m['inst'])
    fun(m['oper'])

print(cpu.strength)
